File: s_GMRES.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.linalg import norm

def arnoldi_iteration(A,V,H,i):
    v = A @ V[:,i]
    for j in range(i):
        H[i,j] = np.dot(V[:,i],v)
        v -= H[i,j]*V[:,i]
    H[j+1,j] = norm(v)
    if H[j+1,j] != 0:
        V[:,j+1] = v/H[j+1,j]
    # M[:,i] = A@v -> this is an extra matrix vector multiply
    # C = S @ M
    # Q,R = QR(C)
    # solve least-squares problem inv(T)(U*g) -> how does this work?
    # x += M*yhat

    return V,H

def gmres(A, b, max_iterations=None, restart=None, tol=1e-6):
    n = A.shape[0]
    if max_iterations is None:
        max_iterations = n
    if restart is None:
        restart = n

    x = np.zeros(n)
    residual = b - A@x
    residual_norm = norm(residual)

    V = np.zeros((n,restart+1))
    H = np.zeros((restart+1,restart))

    converged = False
    iteration = 0

    while not converged and iteration < max_iterations:
        V[:, 0] = residual / residual_norm

        Q, H = arnoldi_iteration(A,V,H,restart)

        e1 = np.zeros(restart+1)
        e1[0] = residual_norm

        y, _, _, _ = np.linalg.lstsq(H,e1)

        x += Q[:,:-1] @ y
        residual = b - A@x
        residual_norm = norm(residual)

        if residual_norm < tol:
            converged = True

        iteration += 1

    return x


# Example usage
A = np.random.randn(10, 10)
b = np.random.randn(10)

x = gmres(A,b)

print("Solution x:")
print(x)

File: s-GMRES.jl
using LinearAlgebra
using Random

# implemented from https://arxiv.org/pdf/2111.00113.pdf
function sGMRES(A,f,x,d,k) # will need to add tolerance
    n = length(f)
    S = randn(2*(d+1),n)
    r = f - A*x
    g = S*r
    B = zeros(n,d)
    B[:,1] = r/norm(r)
    M = zeros(n,d)
    M[:,1] = A*B[:,1]
    M,B = arnoldi(A,M,B,d,k)
    C = S*M
    U,T = qr(C)
    yhat = T \ (U*g)
    xhat = x + M*yhat
end

function arnoldi(A,M,B,d,k)
    for i in 2:d
        w = bSum(i,k,M,B)
        B[:,i] = w/norm(w)
        M[:,i] = A*B[:,i]
    end
end

function bSum(i,k,M,B)
    n = maximum(size(B))
    W = zeros(n,n)
    for j in 1:k
        if i-j > 0
            W -= B[:,i-j]*B[:,i-j]'
        else
            continue
        end
    end
    w = W*M[:,i-1]
    return w
end

# Example usage
n = 10
Random.seed!(123)  # To get consistent results
A = randn(n,n)
f = randn(n)
d = Int(floor(2*n/3))
x = randn(n)
k = n-2

x = sGMRES(A,f,x,d,k)

x_true = A \ f

println("Solution x:")
println(x)
println("True x:")
println(x_true)
println("Error:")
println((x - x_true)/norm(x_true))